From 7807792ca64e2965388583173e12eda2f9aa1bbf Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 10 Jul 2025 17:23:44 +0200
Subject: [PATCH] clk: stm32mp2: fix CA35SS clock controller driver

No more use of_clk_hw_onecell_get/clk_hw_onecell_data as the CA35SS clock
provider is a simple clock provider, without additional parameters in
device tree; the driver use the function of_clk_hw_simple_get.

This patch solves some crash because of_clk_hw_onecell_get parameter (the
third parameter of devm_of_clk_add_hw_provider) is expected to a pointer
to struct clk_init_data (so hw_data) but it is hw : struct clk_hw.

This patch also remove the remaining TODO in the driver.

Fixes: 769bf7d9dff4 ("clk: stm32mp2: add a CA35SS clock controller driver")
Change-Id: Ic04fb553d8f789c7a63d556b35ac034b4aac3f51
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/474109
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Raphael GALLAIS-POU <raphael.gallais-pou@foss.st.com>
Reviewed-by: Raphael GALLAIS-POU <raphael.gallais-pou@foss.st.com>
---
 drivers/clk/stm32/clk-stm32mp2-ca35ss.c | 21 +++++++--------------
 1 file changed, 7 insertions(+), 14 deletions(-)

diff --git a/drivers/clk/stm32/clk-stm32mp2-ca35ss.c b/drivers/clk/stm32/clk-stm32mp2-ca35ss.c
index 89b8fd51a865..cd41bda417ee 100644
--- a/drivers/clk/stm32/clk-stm32mp2-ca35ss.c
+++ b/drivers/clk/stm32/clk-stm32mp2-ca35ss.c
@@ -143,7 +143,12 @@ static unsigned long clk_stm32mp2_ca35ss_recalc_rate(struct clk_hw *hw,
 						     unsigned long parent_rate)
 {
 	u32 rate = 0;
-	(void)smc_recalc_rate(&rate); /* TODO handle status */
+	u32 status;
+
+	status = smc_recalc_rate(&rate);
+	if (status != STM32_SMC_OK)
+		return 0;
+
 	return rate;
 }
 
@@ -187,7 +192,6 @@ static const struct clk_ops clk_stm32mp2_ca35ss_ops = {
 static int clk_stm32mp2_ca35ss_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
-	struct clk_hw_onecell_data *hw_data;
 	struct clk_init_data init = { NULL };
 	struct clk_hw *hw;
 	u32 status, rate;
@@ -199,12 +203,6 @@ static int clk_stm32mp2_ca35ss_probe(struct platform_device *pdev)
 		return -EPERM;
 	}
 
-	hw_data = devm_kzalloc(dev, struct_size(hw_data, hws, 1), GFP_KERNEL);
-	if (!hw_data)
-		return -ENOMEM;
-
-	hw_data->num = 1;
-
 	hw = devm_kzalloc(dev, sizeof(*hw), GFP_KERNEL);
 	if (!hw)
 		return -ENOMEM;
@@ -219,12 +217,7 @@ static int clk_stm32mp2_ca35ss_probe(struct platform_device *pdev)
 		return ret;
 	}
 
-	hw_data->hws[0] = hw;
-
-	platform_set_drvdata(pdev, hw_data);
-
-	ret = devm_of_clk_add_hw_provider(dev, of_clk_hw_onecell_get,
-					  hw);
+	ret = devm_of_clk_add_hw_provider(dev, of_clk_hw_simple_get, hw);
 	if (ret) {
 		dev_err(dev, "Failed to add clock provider\n");
 		clk_hw_unregister(hw);
-- 
2.53.0

